pantheon-bundle/src/main/java/com/redhat/pantheon/model/api/util/SafeResourceTraversal.java

```diff
@@ -0,0 +1,91 @@
+package com.redhat.pantheon.model.api.util;
+
+import com.redhat.pantheon.model.api.Child;
+import com.redhat.pantheon.model.api.Field;
+import com.redhat.pantheon.model.api.SlingModel;
+
+import javax.annotation.Nullable;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static java.util.Optional.ofNullable;
+
+/**
+ * Utility class which offers a null-safe way to traverse deep {@link SlingModel}
+ * structures. Traversals represent paths across multiple nodes ending in deeper
+ * child nodes, or properties. If any of the intermediary nodes is not present, the whole traversal
+ * will still conclude, with a final result of null.<br/>
+ * <br/>
+ * To use it, invoke the {@link SafeResourceTraversal#start(SlingModel)} method with a
+ * model to be traversed. From there on, the whole structure may be traversed all the
+ * way down to specific fields.
+ *
+ * @author Carlos Munoz
+ */
+public class SafeResourceTraversal<T extends SlingModel> implements Supplier<T> {
+
+    private static final SafeResourceTraversal<?> EMPTY = new SafeResourceTraversal<>(null);
```

Comment:
I have a question. Why are we using an unbounded wildcard type for EMPTY instead of `<T>`?

Reply:
never mind. Further down  i see we have genetic types of `<T>` or `<U>` or `<P>`

Reply:
It's unbounded because it's a singleton (private too) and at this point we don't know a specific type parameter to provide. Further down in the `empty()` method you'll see that there is a cast to the specific parameterized type. Since these are all erased types (just for compilation), it all just works.